### ISSUE TITLE
chown fails to set group owner inside container

### ISSUE THREAD

**Issue by pcewing** [moby/moby#40539]
Running `chown` inside of a container to set the group ownership of a file is failing despite having adequate permissions.

In my Dockerfile, I create a user who is a member of two groups. At runtime, that user creates a file and then tries to change the group owner of the file to the user's secondary group, which fails with EPERM.

**Steps to reproduce the issue:**
The steps are a bit long so I've instead provided this gist demonstrating the entire problem:
https://gist.github.com/pcewing/cadaf22b7f6803c8f4afec9439e2719b

**Describe the results you received:**
`chown` fails with EPERM

**Describe the results you expected:**
`chown` successfully changes the ownership of the file.

**Additional information you deem important (e.g. issue happens only occasionally):**
If I remove the `USER` directive from the Dockerfile and instead run `su - john` inside of the container the `chown` call succeeds. This makes me think that `USER` directive may not be applying all of the user's groups correctly.

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           19.03.1
 API version:       1.40
 Go version:        go1.12.5
 Git commit:        74b1e89
 Built:             Thu Jul 25 21:21:22 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.1
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.5
  Git commit:       74b1e89
  Built:            Thu Jul 25 21:19:53 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.6
  GitCommit:        894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc:
  Version:          1.0.0-rc8
  GitCommit:        425e105d5a03fabd737a126ad93d62a9eeede87f
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
Client:
 Debug Mode: false
```

**Output of `docker info`:**

```
Server:
 Containers: 310
  Running: 2
  Paused: 0
  Stopped: 308
 Images: 121
 Server Version: 19.03.1
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc version: 425e105d5a03fabd737a126ad93d62a9eeede87f
 init version: fec3683
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 5.0.0-37-generic
 Operating System: Ubuntu 19.04
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 31.24GiB
 Name: ********
 ID: DYXK:CDRA:USIW:KPUN:JRYB:PX2K:6DUD:OK2H:6AIN:S5BY:MZRX:33DT
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
I'm running on a physical Ubuntu 19.10 host

**Comment by thaJeztah** [comment]
Yes, this is because by default, the container's process is started with only the user's primary group. From [the documentation](https://docs.docker.com/engine/reference/run/#additional-groups)

> By default, the docker container process runs with the supplementary groups looked up for the specified user. If one wants to add more to that list of groups, then one can use this flag:
>
> ```bash
> $ docker run --rm --group-add audio --group-add nogroup --group-add 777 busybox id
> uid=0(root) gid=0(root) groups=10(wheel),29(audio),99(nogroup),777
> ```


Going through your examples:

First, building the image:

```bash
docker build -t myapp -<<'EOF'
FROM centos:latest

RUN groupadd -r -g 1001 john && \
    groupadd -r -g 1050 myapp && \
    useradd -rM -g john -G myapp -u 1001 john && \
    mkdir -p /var/john && \
    chown -R john:john /var/john

USER john:john
EOF
```

As mentioned above, even though the user is member of two groups, when running a docker container, docker does not create a "login" session. When creating a container, docker looks up the numerical IDs (`uid/gid`) of the user, and its _primary_ group (as defined in `/etc/passwd` in the image), and uses those to start the container's process.

You can see this in the output of the `id` command in the example below (adopted from your gist);

```bash
docker container run --rm myapp sh -c 'id && touch /var/john/test.txt && chown john:myapp /var/john/test.txt && ls -la /var/john/test.txt'
# uid=1001(john) gid=1001(john) groups=1001(john)
# chown: changing ownership of '/var/john/test.txt': Operation not permitted
```

If you need additional groups to be set for the container process, you can use the `--group-add` flag. For example, to add the `myapp` group to the process, use `--group-add=myapp`;

```bash
docker container run --rm --group-add=myapp myapp sh -c 'id && touch /var/john/test.txt && chown john:myapp /var/john/test.txt && ls -la /var/john/test.txt'
# uid=1001(john) gid=1001(john) groups=1001(john),1050(myapp)
# -rw-r--r-- 1 john myapp 0 Feb 19 08:13 /var/john/test.txt
```

This time, the output of `id` shows that the process is running with both the `john (1001)` and `myapp (1050)` groups, and the `chown` completes succesfully.


I'm not _exactly_ sure what the reason was for not considering supplemental groups _by default) (perhaps @tianon recalls: the `--group-add` flag was added in https://github.com/moby/moby/pull/10717 to resolve https://github.com/moby/moby/issues/9360 ?)

Reasons I could come up with;

1. Containers designed to be running a single process, and as such were expected the main process to have access to (or primary owner of) files it needs access to (pods and sharing resources were concepts that were implemented later)
2. Complexity; to know about additional groups, docker would have to parse _both_ `/etc/passwd` _and_ `/etc/group` inside the container, and chain together the information.

I would _personally_ consider "2." to be something nice to have "out of the box"; it wouldn't require changes to the [runtime-spec](https://github.com/opencontainers/runtime-spec/blob/v1.0.1/config.md#user), and I _think_ would be possible to implement without breaking backward compatibility; if a supplementary group is also provided through `--group-add`, we can ignore the duplicate. The change in behavior (the supplementary groups being used by default) would give the container more permissions by default, but I wouldn't consider that a security issue, given that this would only be the case if the user was explicitly given access to the group (thus more aligning with expectations).

@tianon @justincormack @AkihiroSuda any thoughts?

**Comment by pcewing** [comment]
Thanks for the thorough response! I guess what is confusing for me is that the `USER` directive functions differently from the `--user` options to `docker run`.

If I change your command above, `id` and `id john` provide different results despite the current user being john:
```
docker container run --rm myapp sh -c 'whoami && id && id john'
# john
# uid=1001(john) gid=1001(john) groups=1001(john)
# uid=1001(john) gid=1001(john) groups=1001(john),1050(myapp)
```

However, if I specify the user as an option to the `run` command, I see the expected results:
```
docker container run --rm --user="john" myapp sh -c 'whoami && id && id john'
# john
# uid=1001(john) gid=1001(john) groups=1001(john),1050(myapp)
# uid=1001(john) gid=1001(john) groups=1001(john),1050(myapp)
```

From that documentation snippet you linked:
> The docker container process runs with the supplementary groups looked up for the specified user

To me this sounds like docker *should* apply the supplementary groups and in some cases it does. The principle of least astonishment would lead most to believe that the Dockerfile `USER` directive would function similarly to the `--user` option. If that's not expected, perhaps the answer is to add something to the Dockerfile reference docs?

**Comment by tianon** [comment]
If you switch from `USER john:john` to `USER john`, you'll see the same (correct) behavior as in `--user=john`:

```console
$ docker run --rm myapp id
uid=1001(john) gid=1001(john) groups=1001(john),1050(myapp)
```

When you specify an explicit group (even if it's the named user's default group), you're communicating to the platform that you want to handle the effective groups of your process yourself.
